### PR TITLE
[eclipse/xtext#1569] Refactor Xtend to Java

### PR DESCRIPTION
**Pull request by kthoms**
Signed-off-by: Karsten Thoms <karsten.thoms@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.builder.standalone/.settings/org.eclipse.jdt.ui.prefs [hunk 1]
```diff
@@ -1,53 +1,60 @@
...
-cleanup.remove_trailing_whitespaces_all=true
-cleanup.remove_trailing_whitespaces_ignore_empty=false
+cleanup.remove_trailing_whitespaces_all=false
+cleanup.remove_trailing_whitespaces_ignore_empty=true
 cleanup.remove_unnecessary_casts=true
 cleanup.remove_unnecessary_nls_tags=true
-cleanup.remove_unused_imports=false
-cleanup.remove_unused_local_variables=false
+cleanup.remove_unused_imports=true
+cleanup.remove_unused_local_variables=true
```
Comment (cdietrich) [inline org.eclipse.xtext.builder.standalone/.settings/org.eclipse.jdt.ui.prefs]: why is this changed?

File: org.eclipse.xtext.common.types.tests/.classpath [hunk 2]
```diff
@@ -43,6 +43,13 @@
 			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
+	<classpathentry including="**/*.java|**/*.xtend" kind="src" output="bin/test" path="xtend-gen">
```
Comment (cdietrich) [inline org.eclipse.xtext.common.types.tests/.classpath]: why is this there now?

Comment (cdietrich) [reply org.eclipse.xtext.common.types.tests/.classpath]: do you have locally uncommited stuff?

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.common.types.descriptions;
+
+import java.util.List;
+
+import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.xbase.lib.Conversions;
```
Comment (cdietrich) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java]: unused import

File: org.eclipse.xtext.purexbase.ide/.classpath [hunk 4]
```diff
@@ -13,6 +13,12 @@
 			<attribute name="ignore_optional_problems" value="true"/>
 		</attributes>
 	</classpathentry>
+	<classpathentry including="**/*.java|**/*.xtend" kind="src" output="bin/main" path="xtend-gen">
```
Comment (cdietrich) [inline org.eclipse.xtext.purexbase.ide/.classpath]: this change is also bad

Comment (cdietrich) [reply org.eclipse.xtext.purexbase.ide/.classpath]: likely old dirty xtend-gen folder locally on your machine

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	protected static class ClassLoaderPackageConfig {
+		private final ClassLoader loader;
+
+		private final boolean bootstrap;
+
+		private final Collection<String> packagePrefixes;
+
+		public ClassLoaderPackageConfig(ClassLoader loader, boolean bootstrap,
+				Collection<String> packagePrefixes) {
+			super();
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: Unneeded

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	private final Provider<ClassGraph> classGraphProvider = ClassGraph::new;
+
+	private final Cache<ClassLoaderPackageConfig, Iterable<ITypeDescriptor>> classLoaderDescriptors = createClassLoaderCache();
+
+	protected Cache<ClassLoaderPackageConfig, Iterable<ITypeDescriptor>> createClassLoaderCache() {
+		return CacheBuilder.newBuilder()
+			.initialCapacity(8)
+			.concurrencyLevel(2)
+			.maximumSize(32)
+			.expireAfterAccess(5,TimeUnit.MINUTES)
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: Space after `5,`

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+			.build();
+	}
+
+	public Iterable<ITypeDescriptor> getDescriptors(ClassLoader classLoader, Collection<String> packagePrefixes) {
+		return this.getDescriptors(classLoader, false, packagePrefixes);
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: No need for `this`

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	protected Iterable<ITypeDescriptor> getDescriptors(ClassLoader classLoader, boolean bootstrap, Collection<String> packagePrefixes) {
+		ClassLoaderPackageConfig clPckCfg = new ClassLoaderPackageConfig(classLoader, bootstrap, packagePrefixes);
+		try {
+			return classLoaderDescriptors.get(clPckCfg, () -> this.loadDescriptors(classLoader, bootstrap, packagePrefixes));
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: No need for `this`

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+			if (e.getCause() instanceof RuntimeException) {
+				throw (RuntimeException) e.getCause();
+			} else {
+				throw new RuntimeException(e.getCause());
+			}
+		}
+	}
+
+	public Iterable<ITypeDescriptor> getBootClasspathDescriptors(Collection<String> packagePrefixes) {
+		return this.getDescriptors(ClassLoader.getSystemClassLoader(), true, packagePrefixes);
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: No need for `this`

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+		ClassGraph classGraph = classGraphProvider.get()
+			.ignoreClassVisibility()
+			.enableClassInfo()
+			.whitelistPackages(packagePrefixes.toArray(new String[packagePrefixes.size()]))
+			.addClassLoader(classLoader);
+		if (bootstrap) {
+			classGraph.enableSystemJarsAndModules();
+		}
+		try (ScanResult scanResult = classGraph.scan()) {
+			return this.loadDescriptors(scanResult);
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.java]: No need for `this`

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.base.Splitter;
+
+public class ClasspathTypeDescriptor implements ITypeDescriptor {
+	private static final Splitter PACKAGE_AND_NESTED_CLASS_SPLITTER = Splitter.onPattern("\\.|\\$");
+
+	private final String name;
+	private final int modifiers;
+
+	public ClasspathTypeDescriptor(String name, int modifiers) {
+		super();
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java]: Unneeded

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		if (other.modifiers != modifiers) {
+			return false;
+		}
+		return true;
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java]: Can't we just use guava's `Objects.hashCode(name)` here?

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+		private final QualifiedName qualifiedName;
+
+		@Override
+		public EClass getEClass() {
+			return null;
+		}
+
+		@Override
+		public EObject getEObjectOrProxy() {
+			throw new UnsupportedOperationException("TODO: auto-generated method stub");
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: Doesn't look like the right comment/exception

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		}
+
+		@Override
+		public URI getEObjectURI() {
+			throw new UnsupportedOperationException("TODO: auto-generated method stub");
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: Doesn't look like the right comment/exception

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+			return null;
+		}
+
+		@Override
+		public String[] getUserDataKeys() {
+			return Strings.EMPTY_ARRAY;
+		}
+
+		public ClassDescription(QualifiedName qualifiedName) {
+			super();
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: No need for this

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+			this.qualifiedName = qualifiedName;
+		}
+
+		@Override
+		public int hashCode() {
+			return 31 * 1 + (qualifiedName == null ? 0 : qualifiedName.hashCode());
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: Use `Objects.hashCode()`?

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+				}
+			} else if (!qualifiedName.equals(other.qualifiedName)) {
+				return false;
+			}
+			return true;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
```
Comment (tivervac) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: I think we can just write this as a String, looks simple enough

**Comment by cdietrich** [comment]
@tivervac @kthoms this happen when you dont remove the xtend-gen folder before calling gradle refresh.
=> gradle will and xtend-gen to classpath as "it is there"

File: org.eclipse.xtext.builder.standalone/.settings/org.eclipse.jdt.ui.prefs (hunk 1 above)
Comment (kthoms) [reply org.eclipse.xtext.builder.standalone/.settings/org.eclipse.jdt.ui.prefs]: These are the cleanup options I have used to clean up the Java sources. I'll remove the file change here.

File: org.eclipse.xtext.common.types.tests/.classpath (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtext.common.types.tests/.classpath]: yes, the xtend-gen folder was still there

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java (hunk 16 above)
Comment (kthoms) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: I think `Objects.hash()` would be better here. `Objects.hashCode()` just calls `hashCode` when the object isn't null. This would not include the field `qualifiedName` to include in the hash operation.

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java (hunk 17 above)
Comment (kthoms) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: The original result would be
```
Foo [
  qualifiedName = foo.bar
]
```
If we don't want to change this it is easier to use ToStringBuilder then writing a string in that format.
I'm tended to leave this.

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java (hunk 16 above)
Comment (tivervac) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/TypeResourceDescription.java]: That's good as well :+1:

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java (hunk 11 above)
Comment (tivervac) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java]: Forgot about this one?

Comment (kthoms) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathTypeDescriptor.java]: Yes, seems so. Will update and merge then.
